File: packages/python/elizaos/bootstrap/services/rolodex.py
"""
Rolodex Service - Contact and relationship management.

This service provides comprehensive contact management capabilities,
including categorization, preferences, and relationship analytics.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import TYPE_CHECKING
from uuid import UUID

from elizaos.types import Service, ServiceType

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime


class ContactCategory(str, Enum):
    """Contact categories."""

    FRIEND = "friend"
    FAMILY = "family"
    COLLEAGUE = "colleague"
    ACQUAINTANCE = "acquaintance"
    VIP = "vip"
    BUSINESS = "business"


@dataclass
class ContactPreferences:
    """Contact preferences."""

    preferred_channel: str | None = None
    timezone: str | None = None
    language: str | None = None
    contact_frequency: str | None = None  # daily, weekly, monthly, quarterly
    do_not_disturb: bool = False
    notes: str | None = None


@dataclass
class ContactInfo:
    """Contact information."""

    entity_id: UUID
    categories: list[str] = field(default_factory=list)
    tags: list[str] = field(default_factory=list)
    preferences: ContactPreferences = field(default_factory=ContactPreferences)
    custom_fields: dict[str, str | int | float | bool] = field(default_factory=dict)
    privacy_level: str = "private"  # public, private, restricted
    last_modified: str = ""


@dataclass
class RelationshipAnalytics:
    """Relationship analytics data."""

    strength: float = 0.0
    interaction_count: int = 0
    last_interaction_at: str | None = None
    average_response_time: float | None = None
    sentiment_score: float | None = None
    topics_discussed: list[str] = field(default_factory=list)


def calculate_relationship_strength(
    interaction_count: int,
    last_interaction_at: str | None = None,
    message_quality: float = 5.0,
    relationship_type: str = "acquaintance",
) -> float:
    """Calculate relationship strength based on interaction patterns."""
    # Base score from interaction count (max 40 points)
    interaction_score = min(interaction_count * 2, 40)

    # Recency score (max 30 points)
    recency_score = 0.0
    if last_interaction_at:
        try:
            last_dt = datetime.fromisoformat(last_interaction_at.replace("Z", "+00:00"))
            days_since = (datetime.now(last_dt.tzinfo) - last_dt).days
            if days_since < 1:
                recency_score = 30
            elif days_since < 7:
                recency_score = 25
            elif days_since < 30:
                recency_score = 15
            elif days_since < 90:
                recency_score = 5
        except Exception:
            pass

    # Quality score (max 20 points)
    quality_score = min(message_quality * 2, 20)

    # Relationship type bonus (max 10 points)
    relationship_bonus = {
        "family": 10,
        "friend": 8,
        "colleague": 6,
        "acquaintance": 4,
        "unknown": 0,
    }

    total = (
        interaction_score
        + recency_score
        + quality_score
        + relationship_bonus.get(relationship_type, 0)
    )
    return max(0.0, min(100.0, round(total, 1)))


class RolodexService(Service):
    """
    Service for managing contacts and relationships.

    Provides capabilities for:
    - Adding and updating contacts
    - Categorizing contacts
    - Tracking relationship analytics
    - Managing contact preferences
    """

    name = "rolodex"
    service_type = ServiceType.UNKNOWN

    @property
    def capability_description(self) -> str:
        """Get the capability description for this service."""
        return "Comprehensive contact and relationship management service"

    def __init__(self) -> None:
        """Initialize the rolodex service."""
        self._contacts: dict[UUID, ContactInfo] = {}
        self._analytics: dict[str, RelationshipAnalytics] = {}
        self._runtime: IAgentRuntime | None = None

    async def start(self, runtime: IAgentRuntime) -> None:
        """Start the rolodex service."""
        self._runtime = runtime
        runtime.logger.info(
            "Rolodex service started",
            src="service:rolodex",
            agentId=str(runtime.agent_id),
        )

    async def stop(self) -> None:
        """Stop the rolodex service."""
        if self._runtime:
            self._runtime.logger.info(
                "Rolodex service stopped",
                src="service:rolodex",
                agentId=str(self._runtime.agent_id),
            )
        self._contacts.clear()
        self._analytics.clear()
        self._runtime = None

    async def add_contact(
        self,
        entity_id: UUID,
        categories: list[str] | None = None,
        preferences: ContactPreferences | None = None,
        custom_fields: dict[str, str | int | float | bool] | None = None,
    ) -> ContactInfo:
        """Add a new contact to the rolodex."""
        contact = ContactInfo(
            entity_id=entity_id,
            categories=categories or ["acquaintance"],
            tags=[],
            preferences=preferences or ContactPreferences(),
            custom_fields=custom_fields or {},
            privacy_level="private",
            last_modified=datetime.utcnow().isoformat(),
        )

        self._contacts[entity_id] = contact

        if self._runtime:
            self._runtime.logger.info(
                f"Added contact {entity_id}",
                src="service:rolodex",
                categories=contact.categories,
            )

        return contact

    async def get_contact(self, entity_id: UUID) -> ContactInfo | None:
        """Get a contact by entity ID."""
        return self._contacts.get(entity_id)

    async def update_contact(
        self,
        entity_id: UUID,
        categories: list[str] | None = None,
        tags: list[str] | None = None,
        preferences: ContactPreferences | None = None,
        custom_fields: dict[str, str | int | float | bool] | None = None,
    ) -> ContactInfo | None:
        """Update an existing contact."""
        contact = self._contacts.get(entity_id)
        if not contact:
            return None

        if categories is not None:
            contact.categories = categories
        if tags is not None:
            contact.tags = tags
        if preferences is not None:
            contact.preferences = preferences
        if custom_fields is not None:
            contact.custom_fields = custom_fields

        contact.last_modified = datetime.utcnow().isoformat()

        return contact

    async def remove_contact(self, entity_id: UUID) -> bool:
        """Remove a contact from the rolodex."""
        if entity_id in self._contacts:
            del self._contacts[entity_id]
            return True
        return False

    async def search_contacts(
        self,
        categories: list[str] | None = None,
        tags: list[str] | None = None,
        search_term: str | None = None,
    ) -> list[ContactInfo]:
        """Search contacts by criteria."""
        results = list(self._contacts.values())

        if categories:
            results = [c for c in results if any(cat in c.categories for cat in categories)]

        if tags:
            results = [c for c in results if any(tag in c.tags for tag in tags)]

        return results

    async def get_all_contacts(self) -> list[ContactInfo]:
        """Get all contacts."""
        return list(self._contacts.values())

    async def get_relationship_analytics(
        self,
        entity_id: UUID,
    ) -> RelationshipAnalytics | None:
        """Get relationship analytics for an entity."""
        key = str(entity_id)
        return self._analytics.get(key)

    async def update_relationship_analytics(
        self,
        entity_id: UUID,
        interaction_count: int | None = None,
        last_interaction_at: str | None = None,
    ) -> RelationshipAnalytics:
        """Update relationship analytics."""
        key = str(entity_id)
        analytics = self._analytics.get(key) or RelationshipAnalytics()

        if interaction_count is not None:
            analytics.interaction_count = interaction_count
        if last_interaction_at is not None:
            analytics.last_interaction_at = last_interaction_at

        # Recalculate strength
        contact = self._contacts.get(entity_id)
        relationship_type = "acquaintance"
        if contact and contact.categories:
            relationship_type = contact.categories[0]

        analytics.strength = calculate_relationship_strength(
            analytics.interaction_count,
            analytics.last_interaction_at,
            relationship_type=relationship_type,
        )

        self._analytics[key] = analytics
        return analytics

File: packages/typescript/src/services/rolodex.ts
import {
  type Entity,
  type IAgentRuntime,
  type JsonValue,
  logger,
  type Metadata,
  type MetadataValue,
  type Relationship,
  Service,
  stringToUuid,
  type UUID,
} from "@elizaos/core";

// Extended Relationship interface with new fields
interface ExtendedRelationship extends Relationship {
  relationshipType?: string;
  strength?: number;
  lastInteractionAt?: string;
  nextFollowUpAt?: string;
}

export interface ContactCategory {
  id: string;
  name: string;
  description?: string;
  color?: string;
}

export interface ContactPreferences {
  preferredCommunicationChannel?: string;
  timezone?: string;
  language?: string;
  contactFrequency?: "daily" | "weekly" | "monthly" | "quarterly";
  doNotDisturb?: boolean;
  notes?: string;
  /** Index signature for metadata compatibility */
  [key: string]: string | boolean | undefined;
}

export interface ContactInfo {
  entityId: UUID;
  categories: string[];
  tags: string[];
  preferences: ContactPreferences;
  customFields: Record<string, JsonValue>;
  privacyLevel: "public" | "private" | "restricted";
  lastModified: string;
}

/** Helper to convert ContactInfo to Metadata for storage */
function contactInfoToMetadata(contactInfo: ContactInfo): Metadata {
  return {
    entityId: contactInfo.entityId,
    categories: contactInfo.categories,
    tags: contactInfo.tags,
    preferences: contactInfo.preferences as MetadataValue,
    customFields: contactInfo.customFields,
    privacyLevel: contactInfo.privacyLevel,
    lastModified: contactInfo.lastModified,
  };
}

/** Helper to convert Metadata back to ContactInfo */
function metadataToContactInfo(data: Metadata): ContactInfo {
  return {
    entityId: data.entityId as UUID,
    categories: data.categories as string[],
    tags: data.tags as string[],
    preferences: data.preferences as ContactPreferences,
    customFields: (data.customFields as Record<string, JsonValue>) ?? {},
    privacyLevel: data.privacyLevel as "public" | "private" | "restricted",
    lastModified: data.lastModified as string,
  };
}

export interface RelationshipAnalytics {
  strength: number;
  interactionCount: number;
  lastInteractionAt?: string;
  averageResponseTime?: number;
  sentimentScore?: number;
  topicsDiscussed: string[];
}

export interface FollowUpSchedule {
  entityId: UUID;
  scheduledAt: string;
  reason: string;
  priority: "high" | "medium" | "low";
  completed: boolean;
  taskId?: UUID;
}

// Entity lifecycle event types
export enum EntityLifecycleEvent {
  CREATED = "entity:created",
  UPDATED = "entity:updated",
  MERGED = "entity:merged",
  RESOLVED = "entity:resolved",
}

export interface EntityEventData {
  entity: Entity;
  previousEntity?: Entity;
  mergedEntities?: Entity[];
  source?: string;
  confidence?: number;
}

/**
 * Calculate relationship strength based on interaction patterns
 */
export function calculateRelationshipStrength({
  interactionCount,
  lastInteractionAt,
  messageQuality = 5,
  relationshipType = "acquaintance",
}: {
  interactionCount: number;
  lastInteractionAt?: string;
  messageQuality?: number;
  relationshipType?: string;
}): number {
  // Base score from interaction count (max 40 points)
  const interactionScore = Math.min(interactionCount * 2, 40);

  // Recency score (max 30 points)
  let recencyScore = 0;
  if (lastInteractionAt) {
    const daysSinceLastInteraction =
      (Date.now() - new Date(lastInteractionAt).getTime()) /
      (1000 * 60 * 60 * 24);
    if (daysSinceLastInteraction < 1) recencyScore = 30;
    else if (daysSinceLastInteraction < 7) recencyScore = 25;
    else if (daysSinceLastInteraction < 30) recencyScore = 20;
    else if (daysSinceLastInteraction < 90) recencyScore = 10;
    else recencyScore = 5;
  }

  // Quality score (max 20 points)
  const qualityScore = (messageQuality / 10) * 20;

  // Relationship type bonus (max 10 points)
  const relationshipBonus: Record<string, number> = {
    family: 10,
    friend: 8,
    colleague: 6,
    acquaintance: 4,
    unknown: 0,
  };

  // Calculate total strength
  const totalStrength =
    interactionScore +
    recencyScore +
    qualityScore +
    (relationshipBonus[relationshipType] ?? 0);

  // Return clamped value between 0 and 100
  return Math.max(0, Math.min(100, Math.round(totalStrength)));
}

export class RolodexService extends Service {
  static serviceType = "rolodex" as const;

  capabilityDescription =
    "Comprehensive contact and relationship management service";

  // In-memory caches for performance
  private contactInfoCache: Map<UUID, ContactInfo> = new Map();
  private analyticsCache: Map<string, RelationshipAnalytics> = new Map();
  private categoriesCache: ContactCategory[] = [];
  private static readonly CONTACT_CACHE_LIMIT = 2000;
  private static readonly ANALYTICS_CACHE_LIMIT = 2000;

  private setCacheWithLimit<K, V>(
    cache: Map<K, V>,
    key: K,
    value: V,
    limit: number,
  ): void {
    if (cache.has(key)) {
      cache.delete(key);
    }
    cache.set(key, value);
    if (cache.size > limit) {
      const firstKey = cache.keys().next().value;
      if (firstKey !== undefined) {
        cache.delete(firstKey);
      }
    }
  }

  async initialize(runtime: IAgentRuntime): Promise<void> {
    this.runtime = runtime;

    // Initialize default categories
    this.categoriesCache = [
      { id: "friend", name: "Friend", color: "#4CAF50" },
      { id: "family", name: "Family", color: "#2196F3" },
      { id: "colleague", name: "Colleague", color: "#FF9800" },
      { id: "acquaintance", name: "Acquaintance", color: "#9E9E9E" },
      { id: "vip", name: "VIP", color: "#9C27B0" },
      { id: "business", name: "Business", color: "#795548" },
    ];

    // Load existing contact info from components
    await this.loadContactInfoFromComponents();

    // Service initialized
    logger.info("[RolodexService] Initialized successfully");
  }

  async stop(): Promise<void> {
    // Clean up caches
    this.contactInfoCache.clear();
    this.analyticsCache.clear();
    this.categoriesCache = [];
    // Service stopped
    logger.info("[RolodexService] Stopped successfully");
  }

  static async start(runtime: IAgentRuntime): Promise<Service> {
    const service = new RolodexService();
    await service.initialize(runtime);
    return service;
  }

  private async loadContactInfoFromComponents(): Promise<void> {
    // Get all rooms for the agent to find entities
    const rooms = await this.runtime.getRooms(
      stringToUuid(`world-${this.runtime.agentId}`),
    );
    const entityIds = new Set<UUID>();

    // Collect unique entity IDs from all rooms
    for (const room of rooms) {
      const entities = await this.runtime.getEntitiesForRoom(room.id, true);
      for (const entity of entities) {
        entityIds.add(entity.id as UUID);
      }
    }

    // Load contact info from components for each entity
    for (const entityId of entityIds) {
      const components = await this.runtime.getComponents(entityId);
      const contactComponent = components.find(
        (c) => c.type === "contact_info" && c.agentId === this.runtime.agentId,
      );

      if (contactComponent?.data) {
        const contactInfo = metadataToContactInfo(
          contactComponent.data as Metadata,
        );
        this.setCacheWithLimit(
          this.contactInfoCache,
          entityId,
          contactInfo,
          RolodexService.CONTACT_CACHE_LIMIT,
        );
      }
    }

    logger.info(
      `[RolodexService] Loaded ${this.contactInfoCache.size} contacts from components`,
    );
  }

  // Contact Management Methods
  async addContact(
    entityId: UUID,
    categories: string[] = ["acquaintance"],
    preferences?: ContactPreferences,
    customFields?: Record<string, JsonValue>,
  ): Promise<ContactInfo> {
    const contactInfo: ContactInfo = {
      entityId,
      categories,
      tags: [],
      preferences: preferences ?? {},
      customFields: customFields ?? ({} as Record<string, JsonValue>),
      privacyLevel: "private",
      lastModified: new Date().toISOString(),
    };

    // Save as component
    await this.runtime.createComponent({
      id: stringToUuid(`contact-${entityId}-${this.runtime.agentId}`),
      type: "contact_info",
      agentId: this.runtime.agentId,
      entityId,
      roomId: stringToUuid(`rolodex-${this.runtime.agentId}`),
      worldId: stringToUuid(`rolodex-world-${this.runtime.agentId}`),
      sourceEntityId: this.runtime.agentId,
      data: contactInfoToMetadata(contactInfo),
      createdAt: Date.now(),
    });

    this.setCacheWithLimit(
      this.contactInfoCache,
      entityId,
      contactInfo,
      RolodexService.CONTACT_CACHE_LIMIT,
    );

    // Emit entity lifecycle event
    const entity = await this.runtime.getEntityById(entityId);
    if (entity) {
      await (
        this.runtime as {
          emitEvent: (
            event: string,
            payload: Record<string, JsonValue | object>,
          ) => Promise<void>;
        }
      ).emitEvent(EntityLifecycleEvent.UPDATED, {
        entityId: entity.id ?? "",
        source: "rolodex",
      });
    }

    logger.info(
      `[RolodexService] Added contact ${entityId} with categories: ${categories.join(", ")}`,
    );
    return contactInfo;
  }

  async updateContact(
    entityId: UUID,
    updates: Partial<ContactInfo>,
  ): Promise<ContactInfo | null> {
    const existing = await this.getContact(entityId);
    if (!existing) {
      logger.warn(`[RolodexService] Contact ${entityId} not found`);
      return null;
    }

    const updated: ContactInfo = {
      ...existing,
      ...updates,
      entityId, // Ensure entityId cannot be changed
      lastModified: new Date().toISOString(),
    };

    // Update component
    const components = await this.runtime.getComponents(entityId);
    const contactComponent = components.find(
      (c) => c.type === "contact_info" && c.agentId === this.runtime.agentId,
    );

    if (contactComponent) {
      await this.runtime.updateComponent({
        ...contactComponent,
        data: contactInfoToMetadata(updated),
      });
    }

    this.setCacheWithLimit(
      this.contactInfoCache,
      entityId,
      updated,
      RolodexService.CONTACT_CACHE_LIMIT,
    );

    logger.info(`[RolodexService] Updated contact ${entityId}`);
    return updated;
  }

  async getContact(entityId: UUID): Promise<ContactInfo | null> {
    // Check cache first
    if (this.contactInfoCache.has(entityId)) {
      const cached = this.contactInfoCache.get(entityId);
      if (cached) {
        return cached;
      }
    }

    // Load from component if not in cache
    const components = await this.runtime.getComponents(entityId);
    const contactComponent = components.find(
      (c) => c.type === "contact_info" && c.agentId === this.runtime.agentId,
    );

    if (contactComponent?.data) {
      const contactInfo = metadataToContactInfo(
        contactComponent.data as Metadata,
      );
      this.setCacheWithLimit(
        this.contactInfoCache,
        entityId,
        contactInfo,
        RolodexService.CONTACT_CACHE_LIMIT,
      );
      return contactInfo;
    }

    return null;
  }

  async removeContact(entityId: UUID): Promise<boolean> {
    const existing = await this.getContact(entityId);
    if (!existing) {
      logger.warn(`[RolodexService] Contact ${entityId} not found`);
      return false;
    }

    // Remove component
    const components = await this.runtime.getComponents(entityId);
    const contactComponent = components.find(
      (c) => c.type === "contact_info" && c.agentId === this.runtime.agentId,
    );

    if (contactComponent) {
      await this.runtime.deleteComponent(contactComponent.id);
    }

    // Remove from cache
    this.contactInfoCache.delete(entityId);

    logger.info(`[RolodexService] Removed contact ${entityId}`);
    return true;
  }

  async searchContacts(criteria: {
    categories?: string[];
    tags?: string[];
    searchTerm?: string;
    privacyLevel?: string;
  }): Promise<ContactInfo[]> {
    const results: ContactInfo[] = [];

    for (const [, contactInfo] of this.contactInfoCache) {
      let matches = true;

      // Check categories
      if (criteria.categories && criteria.categories.length > 0) {
        const categorySet = new Set(contactInfo.categories);
        matches =
          matches && criteria.categories.some((cat) => categorySet.has(cat));
      }

      // Check tags
      if (criteria.tags && criteria.tags.length > 0) {
        const tagSet = new Set(contactInfo.tags);
        matches = matches && criteria.tags.some((tag) => tagSet.has(tag));
      }

      // Check privacy level
      if (criteria.privacyLevel) {
        matches = matches && contactInfo.privacyLevel === criteria.privacyLevel;
      }

      if (matches) {
        results.push(contactInfo);
      }
    }

    // If searchTerm is provided, further filter by entity names
    if (criteria.searchTerm) {
      const searchTermLower = criteria.searchTerm.toLowerCase();
      const entities = await Promise.all(
        results.map((contact) => this.runtime.getEntityById(contact.entityId)),
      );
      const filteredResults: ContactInfo[] = [];
      for (let i = 0; i < results.length; i++) {
        const entity = entities[i];
        if (
          entity?.names.some((name) =>
            name.toLowerCase().includes(searchTermLower),
          )
        ) {
          filteredResults.push(results[i]);
        }
      }
      return filteredResults;
    }

    return results;
  }

  // Relationship Analytics Methods
  async analyzeRelationship(
    sourceEntityId: UUID,
    targetEntityId: UUID,
  ): Promise<RelationshipAnalytics | null> {
    const cacheKey = `${sourceEntityId}-${targetEntityId}`;

    // Check cache first
    if (this.analyticsCache.has(cacheKey)) {
      const cached = this.analyticsCache.get(cacheKey);
      if (cached) {
        // Cache for 1 hour
        if (
          cached.lastInteractionAt &&
          Date.now() - new Date(cached.lastInteractionAt).getTime() < 3600000
        ) {
          return cached;
        }
      }
    }

    // Get relationship
    const relationships = await this.runtime.getRelationships({
      entityId: sourceEntityId,
    });

    const relationship = relationships.find(
      (r) =>
        r.targetEntityId === targetEntityId ||
        r.sourceEntityId === targetEntityId,
    ) as ExtendedRelationship | undefined;

    if (!relationship) {
      return null;
    }

    // Get recent messages between entities
    const messages = await this.runtime.getMemories({
      tableName: "messages",
      entityId: sourceEntityId,
      count: 100,
    });

    const interactions = messages.filter(
      (m) =>
        m.content.inReplyTo === targetEntityId ||
        (m.entityId === targetEntityId &&
          m.content.inReplyTo === sourceEntityId),
    );

    // Calculate metrics
    const interactionCount = interactions.length;
    const lastInteraction = interactions[0];
    const lastInteractionAt = lastInteraction?.createdAt
      ? new Date(lastInteraction.createdAt).toISOString()
      : undefined;

    // Calculate average response time
    let totalResponseTime = 0;
    let responseCount = 0;

    for (let i = 0; i < interactions.length - 1; i++) {
      const current = interactions[i];
      const next = interactions[i + 1];

      if (
        current.entityId !== next.entityId &&
        current.createdAt &&
        next.createdAt
      ) {
        const timeDiff =
          new Date(next.createdAt).getTime() -
          new Date(current.createdAt).getTime();
        totalResponseTime += timeDiff;
        responseCount++;
      }
    }

    const averageResponseTime =
      responseCount > 0 ? totalResponseTime / responseCount : undefined;

    // Extract topics (simplified - could use NLP)
    const topicsSet = new Set<string>();
    for (const msg of interactions) {
      const text = msg.content.text || "";
      // Simple keyword extraction - could be enhanced with NLP
      const keywords = text.match(/\b[A-Z][a-z]+\b/g) || [];
      for (const k of keywords) {
        topicsSet.add(k);
      }
    }

    // Calculate relationship strength
    const strength = calculateRelationshipStrength({
      interactionCount,
      lastInteractionAt,
      relationshipType: relationship.relationshipType,
    });

    const analytics: RelationshipAnalytics = {
      strength,
      interactionCount,
      lastInteractionAt,
      averageResponseTime,
      sentimentScore: 0.7, // Placeholder - could integrate sentiment analysis
      topicsDiscussed: Array.from(topicsSet).slice(0, 10),
    };

    // Update relationship with calculated strength
    if (
      relationship.strength !== strength ||
      relationship.lastInteractionAt !== lastInteractionAt
    ) {
      // Update relationship using components
      const relationshipComponent = {
        id: stringToUuid(`relationship-${relationship.id}`),
        type: "relationship_update",
        agentId: this.runtime.agentId,
        entityId: relationship.sourceEntityId,
        roomId: stringToUuid(`rolodex-${this.runtime.agentId}`),
        worldId: stringToUuid(`rolodex-world-${this.runtime.agentId}`),
        sourceEntityId: relationship.sourceEntityId,
        data: {
          targetEntityId: relationship.targetEntityId,
          strength,
          lastInteractionAt,
          metadata: relationship.metadata,
        } as Metadata,
        createdAt: Date.now(),
      };
      await this.runtime.createComponent(relationshipComponent);
    }

    // Cache the result
    this.setCacheWithLimit(
      this.analyticsCache,
      cacheKey,
      analytics,
      RolodexService.ANALYTICS_CACHE_LIMIT,
    );

    return analytics;
  }

  async getRelationshipInsights(entityId: UUID): Promise<{
    strongestRelationships: Array<{
      entity: Entity;
      analytics: RelationshipAnalytics;
    }>;
    needsAttention: Array<{ entity: Entity; daysSinceContact: number }>;
    recentInteractions: Array<{ entity: Entity; lastInteraction: string }>;
  }> {
    const relationships = await this.runtime.getRelationships({ entityId });
    const insights = {
      strongestRelationships: [] as Array<{
        entity: Entity;
        analytics: RelationshipAnalytics;
      }>,
      needsAttention: [] as Array<{
        entity: Entity;
        daysSinceContact: number;
      }>,
      recentInteractions: [] as Array<{
        entity: Entity;
        lastInteraction: string;
      }>,
    };

    const targets = relationships.map((rel) =>
      rel.sourceEntityId === entityId ? rel.targetEntityId : rel.sourceEntityId,
    );
    const entities = await Promise.all(
      targets.map((target) => this.runtime.getEntityById(target)),
    );
    const analyticsResults = await Promise.all(
      targets.map((target, index) =>
        entities[index] ? this.analyzeRelationship(entityId, target) : null,
      ),
    );

    for (let i = 0; i < relationships.length; i++) {
      const entity = entities[i];
      const analytics = analyticsResults[i];
      if (!entity || !analytics) continue;

      // Strongest relationships
      if (analytics.strength > 70) {
        insights.strongestRelationships.push({ entity, analytics });
      }

      // Needs attention (no contact in 30+ days)
      if (analytics.lastInteractionAt) {
        const daysSince =
          (Date.now() - new Date(analytics.lastInteractionAt).getTime()) /
          (1000 * 60 * 60 * 24);

        if (daysSince > 30) {
          insights.needsAttention.push({
            entity,
            daysSinceContact: Math.round(daysSince),
          });
        }

        // Recent interactions (last 7 days)
        if (daysSince < 7) {
          insights.recentInteractions.push({
            entity,
            lastInteraction: analytics.lastInteractionAt,
          });
        }
      }
    }

    // Sort by relevance
    insights.strongestRelationships.sort(
      (a, b) => b.analytics.strength - a.analytics.strength,
    );
    insights.needsAttention.sort(
      (a, b) => b.daysSinceContact - a.daysSinceContact,
    );
    insights.recentInteractions.sort(
      (a, b) =>
        new Date(b.lastInteraction).getTime() -
        new Date(a.lastInteraction).getTime(),
    );

    return insights;
  }

  // Category Management
  async getCategories(): Promise<ContactCategory[]> {
    return this.categoriesCache;
  }

  async addCategory(category: ContactCategory): Promise<void> {
    if (this.categoriesCache.find((c) => c.id === category.id)) {
      throw new Error(`Category ${category.id} already exists`);
    }

    this.categoriesCache.push(category);
    logger.info(`[RolodexService] Added category: ${category.name}`);
  }

  // Privacy Management
  async setContactPrivacy(
    entityId: UUID,
    privacyLevel: "public" | "private" | "restricted",
  ): Promise<boolean> {
    const contact = await this.getContact(entityId);
    if (!contact) return false;

    contact.privacyLevel = privacyLevel;
    await this.updateContact(entityId, { privacyLevel });

    logger.info(
      `[RolodexService] Set privacy level for ${entityId} to ${privacyLevel}`,
    );
    return true;
  }

  async canAccessContact(
    requestingEntityId: UUID,
    targetEntityId: UUID,
  ): Promise<boolean> {
    const contact = await this.getContact(targetEntityId);
    if (!contact) return false;

    // Agent always has access
    if (requestingEntityId === this.runtime.agentId) return true;

    // Check privacy level
    switch (contact.privacyLevel) {
      case "public":
        return true;
      case "private":
        // Only agent and the entity itself
        return requestingEntityId === targetEntityId;
      case "restricted":
        // Only agent
        return false;
      default:
        return false;
    }
  }
}
